File: crypto/libecc/src/hash/sha3.c
/*
 *  Copyright (C) 2017 - This file is part of libecc project
 *
 *  Authors:
 *      Ryad BENADJILA <ryadbenadjila@gmail.com>
 *      Arnaud EBALARD <arnaud.ebalard@ssi.gouv.fr>
 *      Jean-Pierre FLORI <jean-pierre.flori@ssi.gouv.fr>
 *
 *  Contributors:
 *      Nicolas VIVET <nicolas.vivet@ssi.gouv.fr>
 *      Karim KHALFALLAH <karim.khalfallah@ssi.gouv.fr>
 *
 *  This software is licensed under a dual BSD and GPL v2 license.
 *  See LICENSE file at the root folder of the project.
 */
#include <libecc/lib_ecc_config.h>

#if defined(WITH_HASH_SHA3_224) || defined(WITH_HASH_SHA3_256) || defined(WITH_HASH_SHA3_384) || defined(WITH_HASH_SHA3_512)
#include <libecc/utils/utils.h>
#include <libecc/hash/sha3.h>

/* Init function depending on the digest size. Return 0 on success, -1 on error. */
int _sha3_init(sha3_context *ctx, u8 digest_size)
{
	int ret;

	/*
	 * Check given inpur digest size: we only consider KECCAK versions
	 * mapped on SHA-3 instances (224, 256, 384, 512).
	 */
	MUST_HAVE(((digest_size == (224/8)) || (digest_size == (256/8)) ||
		   (digest_size == (384/8)) || (digest_size == (512/8))), ret, err);
	MUST_HAVE((ctx != NULL), ret, err);

	/* Zeroize the internal state */
	ret = local_memset(ctx->sha3_state, 0, sizeof(ctx->sha3_state)); EG(ret, err);

	ctx->sha3_idx = 0;
	ctx->sha3_digest_size = digest_size;
	ctx->sha3_block_size = (u8)((KECCAK_SLICES * KECCAK_SLICES * sizeof(u64)) - (u8)(2 * digest_size));

	/* Detect endianness */
	ctx->sha3_endian = arch_is_big_endian() ? SHA3_BIG : SHA3_LITTLE;

err:
	return ret;
}

/* Update hash function. Returns 0 on sucess, -1 on error. */
int _sha3_update(sha3_context *ctx, const u8 *input, u32 ilen)
{
	u32 i;
	u8 *state;
	int ret;

	MUST_HAVE(((ctx != NULL) && ((input != NULL) || (ilen == 0))), ret, err);

	state = (u8*)(ctx->sha3_state);

	for (i = 0; i < ilen; i++) {
		u64 idx = (ctx->sha3_endian == SHA3_LITTLE) ? ctx->sha3_idx : SWAP64_Idx(ctx->sha3_idx);
		ctx->sha3_idx++;
		/* Update the state, and adapt endianness order */
		state[idx] ^= input[i];
		if(ctx->sha3_idx == ctx->sha3_block_size){
			KECCAKF(ctx->sha3_state);
			ctx->sha3_idx = 0;
		}
	}

	ret = 0;

err:
	return ret;
}

/* Finalize hash function. Returns 0 on success, -1 on error. */
int _sha3_finalize(sha3_context *ctx, u8 *output)
{
	unsigned int i;
	u8 *state;
	int ret;

	MUST_HAVE((output != NULL) && (ctx != NULL), ret, err);
	MUST_HAVE((ctx->sha3_digest_size <= sizeof(ctx->sha3_state)), ret, err);

	state = (u8*)(ctx->sha3_state);

	/* Proceed with the padding of the last block */
	/* Compute the index depending on the endianness */
	if (ctx->sha3_endian == SHA3_LITTLE) {
		/* Little endian case */
		state[ctx->sha3_idx] ^= 0x06;
		state[ctx->sha3_block_size - 1] ^= 0x80;
	} else {
		/* Big endian case */
		state[SWAP64_Idx(ctx->sha3_idx)] ^= 0x06;
		state[SWAP64_Idx(ctx->sha3_block_size - 1)] ^= 0x80;
	}
	KECCAKF(ctx->sha3_state);
	for(i = 0; i < ctx->sha3_digest_size; i++){
		output[i] = (ctx->sha3_endian == SHA3_LITTLE) ? state[i] : state[SWAP64_Idx(i)];
	}

	ret = 0;

err:
	return ret;
}

#else
/*
 * Dummy definition to avoid the empty translation unit ISO C warning
 */
typedef int dummy;
#endif

File: crypto/libecc/scripts/sha3.py
#/*
# *  Copyright (C) 2017 - This file is part of libecc project
# *
# *  Authors:
# *      Ryad BENADJILA <ryadbenadjila@gmail.com>
# *      Arnaud EBALARD <arnaud.ebalard@ssi.gouv.fr>
# *      Jean-Pierre FLORI <jean-pierre.flori@ssi.gouv.fr>
# *
# *  Contributors:
# *      Nicolas VIVET <nicolas.vivet@ssi.gouv.fr>
# *      Karim KHALFALLAH <karim.khalfallah@ssi.gouv.fr>
# *
# *  This software is licensed under a dual BSD and GPL v2 license.
# *  See LICENSE file at the root folder of the project.
# */
import struct, sys

keccak_rc = [
        0x0000000000000001, 0x0000000000008082, 0x800000000000808A, 0x8000000080008000,
        0x000000000000808B, 0x0000000080000001, 0x8000000080008081, 0x8000000000008009,
        0x000000000000008A, 0x0000000000000088, 0x0000000080008009, 0x000000008000000A,
        0x000000008000808B, 0x800000000000008B, 0x8000000000008089, 0x8000000000008003,
        0x8000000000008002, 0x8000000000000080, 0x000000000000800A, 0x800000008000000A,
        0x8000000080008081, 0x8000000000008080, 0x0000000080000001, 0x8000000080008008
]

keccak_rot = [
        [  0, 36,  3, 41, 18 ],
        [  1, 44, 10, 45,  2 ],
        [ 62,  6, 43, 15, 61 ],
        [ 28, 55, 25, 21, 56 ],
        [ 27, 20, 39,  8, 14 ],
]

def is_python_2():
    if sys.version_info[0] < 3:
        return True
    else:
        return False

# Keccak function
def keccak_rotl(x, l):
	return (((x << l) ^ (x >> (64 - l))) & (2**64-1))

def keccakround(bytestate, rc):
	# Import little endian state
	state = [0] * 25
	for i in range(0, 25):
		to_unpack = ''.join(bytestate[(8*i):(8*i)+8])
		if is_python_2() == False:
			to_unpack = to_unpack.encode('latin-1')
		(state[i],) = struct.unpack('<Q', to_unpack)
	# Proceed with the KECCAK core
	bcd = [0] * 25
	# Theta
	for i in range(0, 5):
		bcd[i] = state[i] ^ state[i + (5*1)] ^ state[i + (5*2)] ^ state[i + (5*3)] ^ state[i + (5*4)] 	
	
	for i in range(0, 5):
		tmp = bcd[(i+4)%5] ^ keccak_rotl(bcd[(i+1)%5], 1)
		for j in range(0, 5):
			state[i + (5 * j)] = state[i + (5 * j)] ^ tmp
	# Rho and Pi
	for i in range(0, 5):
		for j in range(0, 5):
			bcd[j + (5*(((2*i)+(3*j)) % 5))] = keccak_rotl(state[i + (5*j)], keccak_rot[i][j])
	# Chi
	for i in range(0, 5):
		for j in range(0, 5):
			state[i + (5*j)] = bcd[i + (5*j)] ^ (~bcd[((i+1)%5) + (5*j)] & bcd[((i+2)%5) + (5*j)])
	# Iota
	state[0] = state[0] ^ keccak_rc[rc]
	# Pack the output state
	output = [0] * (25 * 8)
	for i in range(0, 25):
		packed = struct.pack('<Q', state[i])
		if is_python_2() == True:
			output[(8*i):(8*i)+1] = packed
		else:
			output[(8*i):(8*i)+1] = packed.decode('latin-1')
	return output

def keccakf(bytestate):
	for rnd in range(0, 24):
		bytestate = keccakround(bytestate, rnd)	
	return bytestate

# SHA-3 context class
class Sha3_ctx(object):
	def __init__(self, digest_size):
		self.digest_size = digest_size / 8
		self.block_size = (25*8) - (2 * (digest_size / 8))
		self.idx = 0
		self.state = [chr(0)] * (25 * 8)
	def digest_size(self):
		return self.digest_size
	def block_size(self):
		return self.block_size
	def update(self, message):
		if (is_python_2() == False):
			message = message.decode('latin-1')
		for i in range(0, len(message)):
			self.state[self.idx] = chr(ord(self.state[self.idx]) ^ ord(message[i]))
			self.idx = self.idx + 1
			if (self.idx == self.block_size):
				self.state = keccakf(self.state)
				self.idx = 0
	def digest(self):
		self.state[self.idx] = chr(ord(self.state[self.idx]) ^ 0x06)
		self.state[int(self.block_size - 1)] = chr(ord(self.state[int(self.block_size - 1)]) ^ 0x80)
		self.state = keccakf(self.state)
		digest = ''.join(self.state[:int(self.digest_size)])
		if is_python_2() == False:
			digest = digest.encode('latin-1')
		return digest
